File: src/appl.f/zeroin.f
c
c      a zero of the function  f(x)  is computed in the interval ax,bx .
c
c  input..
c
c  ax     left endpoint of initial interval
c  bx     right endpoint of initial interval
c  f      function subprogram which evaluates f(x) for any x in
c         the interval  ax,bx
c  tol    desired length of the interval of uncertainty of the
c         final result ( .ge. 0.0d0)
c
c
c  output..
c
c  zeroin abcissa approximating a zero of  f  in the interval ax,bx
c
c
c      it is assumed  that   f(ax)   and   f(bx)   have  opposite  signs
c  without  a  check.  zeroin  returns a zero  x  in the given interval
c  ax,bx  to within a tolerance  4*macheps*abs(x) + tol, where macheps
c  is the relative machine precision.
c      this function subprogram is a slightly  modified  translation  of
c  the algol 60 procedure  zero  given in  richard brent, algorithms for
c  minimization without derivatives, prentice - hall, inc. (1973).
c
c

      double precision function zeroin(ax,bx,f,tol)
c     implicit undefined(a-z)
      double precision ax,bx,f,tol
      double precision  a,b,c,d,e,eps,fa,fb,fc,tol1,xm,p,q,r,s
      double precision  dabs,dsign
c
c  compute eps, the relative machine precision
c
      eps = 1.0d0
   10 eps = eps/2.0d0
      tol1 = 1.0d0 + eps
      if (tol1 .gt. 1.0d0) go to 10
c
c initialization
c
      a = ax
      b = bx
      fa = f(a)
      fb = f(b)
c
c begin step
c
   20 c = a
      fc = fa
      d = b - a
      e = d
   30 if (dabs(fc) .ge. dabs(fb)) go to 40
      a = b
      b = c
      c = a
      fa = fb
      fb = fc
      fc = fa
c
c convergence test
c
   40 tol1 = 2.0d0*eps*dabs(b) + 0.5d0*tol
      xm = .5d0*(c - b)
      if (dabs(xm) .le. tol1) go to 90
      if (fb .eq. 0.0d0) go to 90
c
c is bisection necessary
c
      if (dabs(e) .lt. tol1) go to 70
      if (dabs(fa) .le. dabs(fb)) go to 70
c
c is quadratic interpolation possible
c
      if (a .ne. c) go to 50
c
c linear interpolation
c
      s = fb/fa
      p = 2.0d0*xm*s
      q = 1.0d0 - s
      go to 60
c
c inverse quadratic interpolation
c
   50 q = fa/fc
      r = fb/fc
      s = fb/fa
      p = s*(2.0d0*xm*q*(q - r) - (b - a)*(r - 1.0d0))
      q = (q - 1.0d0)*(r - 1.0d0)*(s - 1.0d0)
c
c adjust signs
c
   60 if (p .gt. 0.0d0) q = -q
      p = dabs(p)
c
c is interpolation acceptable
c
      if ((2.0d0*p) .ge. (3.0d0*xm*q - dabs(tol1*q))) go to 70
      if (p .ge. dabs(0.5d0*e*q)) go to 70
      e = d
      d = p/q
      go to 80
c
c bisection
c
   70 d = xm
      e = d
c
c complete step
c
   80 a = b
      fa = fb
      if (dabs(d) .gt. tol1) b = b + d
      if (dabs(d) .le. tol1) b = b + dsign(tol1, xm)
      fb = f(b)
      if ((fb*(fc/dabs(fc))) .gt. 0.0d0) go to 20
      go to 30
c
c done
c
   90 zeroin = b
      return
      end

      subroutine root1d(ax, bx, f, tol, root)
c     implicit undefined(a-z)
      double precision ax, bx, f, tol, root, zeroin
      external f
      if(ax*bx .lt. 0) root = zeroin(ax, bx, f, tol)
      return
      end

File: src/appl.c/zeroin.c
/*
 *      a zero of the function  f(x)  is computed in the interval ax,bx .
 *
 *  input..
 *
 *  ax     left endpoint of initial interval
 *  bx     right endpoint of initial interval
 *  f      function subprogram which evaluates f(x) for any x in
 *         the interval  ax,bx
 *  tol    desired length of the interval of uncertainty of the
 *         final result ( .ge. 0.0d0)
 *
 *
 *  output..
 *
 *  zeroin abcissa approximating a zero of  f  in the interval ax,bx
 *
 *
 *      it is assumed  that   f(ax)   and   f(bx)   have  opposite  signs
 *  without  a  check.  zeroin  returns a zero  x  in the given interval
 *  ax,bx  to within a tolerance  4*macheps*abs(x) + tol, where macheps
 *  is the relative machine precision.
 *      this function subprogram is a slightly  modified  translation  of
 *  the algol 60 procedure  zero  given in  richard brent, algorithms for
 *  minimization without derivatives, prentice - hall, inc. (1973).
 */


#include "Fortran.h"

double F77_SYMBOL(zeroin) (double *ax, double *bx, double (*f) (),
				   double *tol) {
	double ret_val, d__1;
	double a, b, c, d, e, p, q, r, s, fa, fb, fc, xm, eps, tol1;

	/* compute eps, the relative machine precision */

	eps = 1.;
L10:
	eps /= 2.;
	tol1 = eps + 1.;
	if (tol1 > 1.) {
		goto L10;
	}
	/* initialization */

	a = *ax;
	b = *bx;
	fa = (*f) (&a);
	fb = (*f) (&b);

	/* begin step */

L20:
	c = a;
	fc = fa;
	d = b - a;
	e = d;
L30:
	if (abs(fc) >= abs(fb)) {
		goto L40;
	}
	a = b;
	b = c;
	c = a;
	fa = fb;
	fb = fc;
	fc = fa;

	/* convergence test */

L40:
	tol1 = eps * 2. * abs(b) + *tol * .5;
	xm = (c - b) * .5;
	if (abs(xm) <= tol1) {
		goto L90;
	}
	if (fb == 0.) {
		goto L90;
	}
	/* is bisection necessary */

	if (abs(e) < tol1) {
		goto L70;
	}
	if (abs(fa) <= abs(fb)) {
		goto L70;
	}
	/* is quadratic interpolation possible */

	if (a != c) {
		goto L50;
	}
	/* linear interpolation */

	s = fb / fa;
	p = xm * 2. * s;
	q = 1. - s;
	goto L60;

	/* inverse quadratic interpolation */

L50:
	q = fa / fc;
	r = fb / fc;
	s = fb / fa;
	p = s * (xm * 2. * q * (q - r) - (b - a) * (r - 1.));
	q = (q - 1.) * (r - 1.) * (s - 1.);

	/* adjust signs */

L60:
	if (p > 0.) {
		q = -q;
	}
	p = abs(p);

	/* is interpolation acceptable */

	if (p * 2. >= xm * 3. * q - (d__1 = tol1 * q, abs(d__1))) {
		goto L70;
	}
	if (p >= (d__1 = e * .5 * q, abs(d__1))) {
		goto L70;
	}
	e = d;
	d = p / q;
	goto L80;

	/* bisection */

L70:
	d = xm;
	e = d;

	/* complete step */

L80:
	a = b;
	fa = fb;
	if (abs(d) > tol1) {
		b += d;
	}
	if (abs(d) <= tol1) {
		b += DSIGN(&tol1, &xm);
	}
	fb = (*f) (&b);
	if (fb * (fc / abs(fc)) > 0.) {
		goto L20;
	}
	goto L30;

	/* done */

L90:
	ret_val = b;
	return ret_val;
}

int 
F77_SYMBOL(root1d) (double *ax, double *bx, double (*f) (),
	double *tol, double *root)
{
	extern double F77_SYMBOL(zeroin) ();

	if (*ax * *bx < 0.) {
		*root = F77_SYMBOL(zeroin)(ax, bx, f, tol);
	}
	return 0;
}
